pallets/nfts/src/features/approvals.rs

```diff
@@ -165,10 +179,197 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+			collection,
+			item: None,
+			owner: origin,
+			delegate,
+			deadline,
+		});
+
+		Ok(())
+	}
+
+	/// Cancels the transfer of items in the collection that owned by the origin to a delegate.
+	///
+	/// This function is used to cancel the approval for the transfer of items in the `collection`
+	/// that owned by the `origin` to a `delegate`. After canceling the approval, the function
+	/// returns the `origin` back the deposited fund and emits the `ApprovalCancelled` event.
+	///
+	/// - `origin`: The account grants permission to cancel the transfer.
+	/// - `collection`: The identifier of the collection.
+	/// - `delegate`: The account that was previously allowed to take control of items in the
+	///   collection that owned by the origin.
+	pub(crate) fn do_cancel_collection_approval(
+		origin: T::AccountId,
+		collection: T::CollectionId,
+		delegate: T::AccountId,
+	) -> DispatchResult {
+		let (_, deposit) = CollectionApprovals::<T, I>::take((&collection, &origin, &delegate))
+			.ok_or(Error::<T, I>::UnknownCollection)?;
+
+		T::Currency::unreserve(&origin, deposit);
+
+		Self::deposit_event(Event::ApprovalCancelled {
+			collection,
+			owner: origin,
+			item: None,
+			delegate,
+		});
+
+		Ok(())
+	}
+
+	/// Clears all collection approvals.
+	///
+	/// This function is used to clear all approvals to transfer items in the `collection` that
+	/// owned by the `origin` to a `delegate`. After clearing all approvals, the function returns
+	/// the `origin` back the deposited fund of each collection approval and emits the
+	/// `AllApprovalsCancelled` event.
+	///
+	/// - `origin`: The account grants permission to clear the transfer.
+	/// - `collection`: The collection ID containing the item.
+	/// - `witness_approvals`: Information on the collection approvals cleared. This must be
+	///   correct.
+	pub(crate) fn do_clear_all_collection_approvals(
+		origin: T::AccountId,
+		collection: T::CollectionId,
+		witness_approvals: u32,
+	) -> DispatchResult {
+		let mut removed_approvals: u32 = 0;
+		// Iterate and remove each collection approval, return the deposited fund back to the
+		// `origin`.
+		for (_, (_, deposit)) in CollectionApprovals::<T, I>::drain_prefix((collection, &origin)) {
+			ensure!(removed_approvals < witness_approvals, Error::<T, I>::BadWitness);
+			T::Currency::unreserve(&origin, deposit);
+			removed_approvals.saturating_inc();
+		}
+		ensure!(removed_approvals == witness_approvals, Error::<T, I>::BadWitness);
+		Self::deposit_event(Event::AllApprovalsCancelled { collection, item: None, owner: origin });
+		Ok(())
+	}
+
+	/// Checks whether the `delegate` is approved to transfer items in the `collection` that owned
+	/// by the `account`.
+	///
+	/// - `collection`: The identifier of the collection.
+	/// - `account`: The account that granted the permission for `delegate` to transfer items.
+	/// - `delegate`: The account that was previously allowed to take control of items in the
```

Comment:
Sloppy, please read the comments you copy paste and read your comments when you self review.